File: 2020/src/day07.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::str::SplitWhitespace;

fn bags_inside(current: usize, bags: &Vec<Vec<(usize, usize)>>) -> usize {
    let mut tot = 0;
    for (count, i) in &bags[current] {
        tot += count * (1 + bags_inside(*i, bags));
    }

    tot
}

fn get_color(words: &mut SplitWhitespace) -> String {
    words.next().unwrap().to_owned() + words.next().unwrap()
}

pub fn solve(input: String) -> String {
    let mut colors = HashMap::new();
    let lines: Vec<_> = input.lines().collect();
    let mut outer: Vec<Vec<usize>> = Vec::new();
    let mut inner: Vec<Vec<(usize, usize)>> = Vec::new();

    for line in &lines {
        let mut words = line.split_whitespace();
        colors.insert(get_color(&mut words), colors.len());
        outer.push(Vec::new());
        inner.push(Vec::new());
    }

    for line in &lines {
        let mut words = line.split_whitespace();
        let o = colors[&get_color(&mut words)];

        let mut n = words.nth(2);
        while n.is_some() && n.unwrap() != "no" {
            let i = colors[&get_color(&mut words)];
            let count: usize = n.unwrap().parse().unwrap();
            outer[i].push(o);
            inner[o].push((count, i));
            n = words.nth(1);
        }
    }

    let mut to_visit = vec![colors["shinygold"]];
    let mut visited: HashSet<_> = vec![colors["shinygold"]].into_iter().collect();

    while let Some(current) = to_visit.pop() {
        for o in &outer[current] {
            if !visited.contains(o) {
                visited.insert(*o);
                to_visit.push(*o);
            }
        }
    }

    let ans1 = visited.len() - 1;
    let ans2 = bags_inside(colors["shinygold"], &inner);

    format!("{}, {}", ans1, ans2)
}

File: 2025/src/day07.zig
const std = @import("std");
const lib = @import("lib.zig");

// low 5224

pub fn solve(input: []const u8) !struct { []u8, []u8 } {
    var ans1: i64 = 0;
    var ans2: i64 = 0;
    var lines: std.ArrayList([]const u8) = .empty;
    var counts: std.ArrayList([]i64) = .empty;
    var it = lib.lines(input);
    while (it.next()) |line| {
        try lines.append(lib.a, line);
        try counts.append(lib.a, try lib.a.alloc(i64, line.len));
        var i: usize = 0;
        for (line) |c| {
            if (c == 'S') {
                counts.items[counts.items.len - 1][i] = 1;
            } else {
                counts.items[counts.items.len - 1][i] = 0;
            }
            i += 1;
        }
    }
    var y: usize = 0;
    for (lines.items) |l| {
        if (y == 0) {
            y += 1;
            continue;
        }
        var x: usize = 0;
        for (l) |c| {
            if (c == '.') {
                // std.debug.print("{}\n", .{counts.items[y][x]});
                // std.debug.print("{}\n", .{counts.items[y - 1][x]});
                counts.items[y][x] += counts.items[y - 1][x];
            } else if (c == '^') {
                if (counts.items[y - 1][x] > 0) {
                    ans1 += 1;
                    counts.items[y][x - 1] += counts.items[y - 1][x];
                    counts.items[y][x + 1] += counts.items[y - 1][x];
                }
            }
            x += 1;
        }
        y += 1;
    }

    for (counts.items[counts.items.len - 1]) |c| {
        ans2 += c;
    }

    return .{ try lib.intToString(ans1), try lib.intToString(ans2) };
}
